### PR TITLE
Content metadata fields, API, and Search

### LINKED ISSUE #8311: Content Metadata for Categorical search
Ability to store new categorical metadata fields in the database, for filtering, and search.

### LINKED ISSUE #8312: Update ContentNode API for categorical search metadata
Update ContentNode API to return relevant categorical search metadata

### LINKED ISSUE #8313: Update search backend for categorical metadata
Need to be able to leverage new fields for search.

### LINKED ISSUE #8493: Keyword search - Searching by a single letter causes a 500 error
## Observed behavior
Searching by a single letter causes a 500 error

## Expected behavior
The search should function correctly and return results.

## Steps to reproduce the issue
1. Install the following [build](https://buildkite.com/learningequality/kolibri-python-package/builds/3846) 
2. Go to Learn>Library
3. Enter a search term consisting of a single character

## Usage Details
 - OS: Windows 10 and Ubuntu
 - Browser: Chrome

### PR DESCRIPTION
**Pull request by rtibbles**
## Summary
* Adds backend fields for new metadata fields
* Updates content schema and tests for content import
* Adds upgrade migration for content kinds to learning activities for those that can be automatically mapped
* Adds import migrations for content kinds to learning activities when importing from older database schemas
* Updates ContentNode API endpoints to return the new metadata as arrays of ids
* Adds ValuesViewset compatible cursor pagination for use with the ContentNode endpoint
* Updates LimitOffset pagination to fix some errors in previous implementation
* Adds filtering of the ContentNode endpoint by the new metadata fields, using `contains` queries
* Adds a new 'keyword' filter that implements a simpler version of the behaviour of the ContentNodeSearchViewset
* Integrates the Library Search interface with backend filtering
* Cleans up and deletes a lot of previous frontend Search code and defers all searching to the Library page
* Adds annotation and caching of labels that are available for queries
* Bootstraps available labels for all available nodes into the learn page
* Uses bootstrapped available labels to restrict the filters that are shown to users in search (off by default in development for now)
* Returns available labels for the query when paginated requests are made to the ContentNode endpoint
* Uses available labels from a search to disable filter options

## References
Fixes #8311 
Fixes #8312 
Fixes #8313
Fixes #8493

## Reviewer guidance
Caveat - the introduction of pure ids into this PR means that areas of the frontend code that were relying on english string ids coming from the backend are now broken.

Test the Library page and confirm that searching is working as might be expected.

1. Filtering by Learning Activity - ensure that filtering by Learning Activity does indeed produce a list of content for just that learning activity (although rendering of both the content and the icons for the learning activities will currently be broken).
2. Filtering by channel - ensure that filtering by channel produces results from within that channel only
3. Filtering by language - ensure that filtering by languages produces search results only for that language
4. Filter by keyword search - ensure that filtering by keyword searches produces somewhat sensible results (plenty of room for interpretation here, but compare against previous keyword search behaviour!)

In all of the above cases, ensuring that the offered filter options match the content that has been loaded onto the device.

Running from production assets specific:
* Ensure that only filters that are present on the local device metadata (i.e. learning activities, language, channels) appear to the end user.

Things that are definitely broken:
* Rendering content
* Most card metadata display

search

----

## Testing checklist

- [x] Contributor has fully tested the PR manually
- [x] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [ ] Critical and brittle code paths are covered by unit tests

## PR process

- [x] PR has the correct target branch and milestone
- [x] PR has 'needs review' or 'work-in-progress' label
- [x] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

## Reviewer checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

### CODE REVIEW

**Comment by pcenov** [comment]
Tested this one today in Win 10 and Ubuntu. Here's the test report:

<table>
<tr>
    <td> Scenario
	<td> Result
<tr>
	<td> Filtering by Learning Activity
	<td> Pass
<tr>
	<td> Filtering by channel
	<td> Pass
<tr>
	<td> Filtering by language
	<td> Pass
<tr>
	<td> Filtering by keyword search
	<td> Pass
<tr>
	<td> Ensure that only filters that are present on the local device metadata appear
	<td> Pass
</table>

I've reported separately the following issue:

|Searching by a single letter results in a 500 error||https://github.com/learningequality/kolibri/issues/8493|
|--|--|--|

**Comment by rtibbles** [comment]
Thanks @pcenov I've updated the PR to fix the issue, I believe.

**Review by marcellamaki — COMMENTED** [review]
Front end code read through looks good, and manual QA looks good. I've also tested the fix you made after @pcenov's review and see results and no 500 errors! Happy to approve and merge first things tomorrow so I can start resolving merge conflicts, as long as @jredrejo feels there are no blocking concerns on the backend 🎉

File: kolibri/core/content/base_models.py [hunk 1]
```diff
@@ -132,6 +132,17 @@ class ContentNode(MPTTModel):
...
+    # Fields for metadata labels
+    grade_levels = models.TextField(blank=True, null=True)
+    resource_types = models.TextField(blank=True, null=True)
+    learning_activities = models.TextField(blank=True, null=True)
+    accessibility_labels = models.TextField(blank=True, null=True)
+    categories = models.TextField(blank=True, null=True)
+    learner_needs = models.TextField(blank=True, null=True)
+
+    # The (suggested) duration of a resource, in seconds.
+    duration = models.PositiveIntegerField(null=True, blank=True)
```
Comment (jredrejo) [inline kolibri/core/content/base_models.py:144]: I don't understand what is the purpose of this field in the context of this PR

File: kolibri/core/utils/pagination.py [hunk 2]
```diff
@@ -156,7 +159,7 @@ def paginate_queryset(self, queryset, request, view=None):
         # Ensure we evaluate the sliced queryset and do an explicit filter by the subsequent PKs
         # to avoid issues when we later try to annotate this queryset.
         return queryset.filter(
-            id__in=list(
+            pk__in=list(
                 queryset.values_list("pk", flat=True).distinct()[
```
Comment (jredrejo) [inline kolibri/core/utils/pagination.py:163]: When giving the output, its giving distinct node ids, but that shows to the user repeated resources. wouldn't it be more user friendly to filter by distinct content_id?
For the user it's the same resource, no matter where it's located in the channel tree.

**Review by jredrejo — COMMENTED** [review]
It's probably a frontend issue, but saying he results are 25 and having  the `show more `button is not very intuitive. At first I thought the search only gave 25 results in total. It would be good to know how many results gives a query, to tune it. (unless this makes a great impact on the search performance, counting can be expensive)

**Comment by marcellamaki** [comment]
> It's probably a frontend issue, but saying he results are 25 and having the `show more `button is not very intuitive. At first I thought the search only gave 25 results in total. It would be good to know how many results gives a query, to tune it. (unless this makes a great impact on the search performance, counting can be expensive)

Good point @jredrejo. Maybe when I make the other front end changes after merge, I could have it be updated to have the full number of results in the title (i.e. 27 results) but then have the pagination which would show the first 25 results, and then the show more button? Would that work with how the data is structured now, @rtibbles?

**Comment by rtibbles** [comment]
> Good point @jredrejo. Maybe when I make the other front end changes after merge, I could have it be updated to have the full number of results in the title (i.e. 27 results) but then have the pagination which would show the first 25 results, and then the show more button? Would that work with how the data is structured now, @rtibbles?

Yes, as @jredrejo pointed out knowing the full number in advance will impact performance, so I have not included it (it is using cursor based pagination that does not include this data).

Unfortunately the string that would have allowed better rendering of this 'More than X results' has an issue, as the variable in it is not formatted as a number.

We can also do an update to this string, as it is only for the ICU syntax portion and do a find replace on that to allow that string to be used instead. I had done this as a workaround.

File: kolibri/core/content/base_models.py (hunk 1 above)
Comment (rtibbles) [reply kolibri/core/content/base_models.py:144]: It's not directly related - except that it's a new metadata field that we are adding in the course of 0.15, so I added them all at once.

File: kolibri/core/utils/pagination.py (hunk 2 above)
Comment (rtibbles) [reply kolibri/core/utils/pagination.py:163]: Yes, can update - but will do so in the specific subclass for ContentNode!

**Comment by pcenov** [comment]
Hi @rtibbles, I confirm that https://github.com/learningequality/kolibri/issues/8493 is fixed now.

File: kolibri/core/utils/pagination.py (hunk 2 above)
Comment (rtibbles) [reply kolibri/core/utils/pagination.py:163]: We discussed this slightly further - have filed a follow up issue for us to look into solutions here: https://github.com/learningequality/kolibri/issues/8498

**Review by jredrejo — APPROVED** [review]
After the creation of #8498 everything looks fine for me in this PR

**Pull request merged**